components/global/core/io.cellery.observability.telemetry.deduplicator/src/main/java/io/cellery/observability/telemetry/deduplicator/DeduplicationWindowExtension.java

```diff
@@ -0,0 +1,403 @@
+        String oldEventDestinationComponent = (String) destinationComponentExecutor.execute(oldEvent);
+
+        String currentEventTraceId = (String) traceIdExecutor.execute(currentEvent);
+        String currentEventSpanId = (String) spanIdExecutor.execute(currentEvent);
+        String currentEventParentSpanId = (String) parentSpanIdExecutor.execute(currentEvent);
+        String currentEventSourceNamespace = (String) sourceNamespaceExecutor.execute(currentEvent);
+        String currentEventSourceInstance = (String) sourceInstanceExecutor.execute(currentEvent);
+        String currentEventSourceComponent = (String) sourceComponentExecutor.execute(currentEvent);
+        String currentEventDestinationNamespace = (String) destinationNamespaceExecutor.execute(currentEvent);
+        String currentEventDestinationInstance = (String) destinationInstanceExecutor.execute(currentEvent);
+        String currentEventDestinationComponent = (String) destinationComponentExecutor.execute(currentEvent);
+
+        return currentEventSourceNamespace.equals(oldEventSourceNamespace)
+                && currentEventSourceInstance.equals(oldEventSourceInstance)
+                && currentEventSourceComponent.equals(oldEventSourceComponent)
+                && currentEventDestinationNamespace.equals(oldEventDestinationNamespace)
+                && currentEventDestinationInstance.equals(oldEventDestinationInstance)
+                && currentEventDestinationComponent.equals(oldEventDestinationComponent)
+                && oldEventTraceId.equals(currentEventTraceId)
+                && (oldEventSpanId.equals(currentEventParentSpanId) || oldEventParentSpanId.equals(currentEventSpanId));
+    }
+
+    private StreamEvent mergeEvents(StreamEvent oldEvent,
+            StreamEvent currentEvent) {
+        Long currentEventRequestSizeBytes = (Long) requestSizeBytesExecutor.execute(currentEvent);
+        Long currentEventResponseDurationNanoSec = (Long) responseDurationNanoSecExecutor.execute(currentEvent);
+        Long currentEventResponseSizeBytes = (Long) responseSizeBytesExecutor.execute(currentEvent);
+        Long oldEventRequestSizeBytes = (Long) requestSizeBytesExecutor.execute(oldEvent);
+        Long oldEventResponseDurationNanoSec = (Long) responseDurationNanoSecExecutor.execute(oldEvent);
+        Long oldEventResponseSizeBytes = (Long) responseSizeBytesExecutor.execute(oldEvent);
+
+        Long maxRequestSizeBytes = Long.max(oldEventRequestSizeBytes, currentEventRequestSizeBytes);
+        Long maxResponseDurationNanoSec = Long.max(oldEventResponseDurationNanoSec,
+                currentEventResponseDurationNanoSec);
+        Long maxResponseSizeBytes = Long.max(oldEventResponseSizeBytes, currentEventResponseSizeBytes);
+
+        StreamEvent mergedEvent = new StreamEvent(0, 0,
+                 Constants.OUTPUT_DATA_SIZE);
+        Object[] currentOutputs = currentEvent.getOutputData();
```

Comment:
Ideally we shouldn't be doing this. updating the Event in place will make it impossible to update the stream definition of the incoming event in the Siddhi App without changing the stream processor code. This will be harder to maintain.

What we should be doing is, returning new three attributes with the merged values. In the Siddhi app we should select those attributes properly.